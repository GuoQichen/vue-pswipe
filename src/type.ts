import Vue from 'vue'
import { Options, Item } from 'photoswipe/dist/photoswipe-ui-default'

/**
 * v-pswp directive config options
 */
export interface PswpDirectiveOptions {
    /**
     * path to image
     */
    src: string
    /**
     * image size, 'width x height', eg: '100x100'
     */
    size?: string
    /**
     * small image placeholder,
     * main (large) image loads on top of it,
     * if you skip this parameter - grey rectangle will be displayed,
     * try to define this property only when small image was loaded before
     */
    msrc?: string
    /**
     * used by Default PhotoSwipe UI
     * if you skip it, there won't be any caption
     */
    title?: string
    /**
     * to make URLs to a single image look like this: http://example.com/#&gid=1&pid=custom-first-id
     * instead of: http://example.com/#&gid=1&pid=1
     * enable options history:true, galleryPIDs:true and add pid (unique picture identifier)
     */
    pid?: string | number
}

export interface PswpItem extends Item {
    el: HTMLElement
    src: string
    msrc?: string
    pid?: number | string
    verticalRotated?: boolean
}

export interface ManualImgItem extends Partial<Item> {
    src: string
}

export type ManualHtmlItem = Partial<Item> 

export interface CurrentPswpItem extends PswpItem {
    container: HTMLElement
    loaded: boolean
}

export type PswpOptions = Options

export interface OpenPhotoSwipeArgs {
    index: number | string
    fromURL?: boolean
    thumbEls?: HTMLElement[]
}

export interface BeforeOpenEvent {
    index: number
    target: HTMLElement
    items: PswpItem[]
    options: PswpOptions
}

export interface Size {
    w: number
    h: number
}

export interface ManualCreateArgs {
    items: ManualOpenItem[]
    options?: PswpOptions
}

interface CreatePhotoSwipeArgs extends ManualCreateArgs {
    context?: Vue
}

export interface PswpProps {
    options: PswpOptions
    auto: boolean
    bubble: boolean
    lazy: boolean
    filter: Function
    rotate: boolean
}

// types
export type Fn = (...args: any[]) => void

export type Pswp = PhotoSwipe<PswpOptions>

export type BeforeOpen = (continued?: boolean) => void

export type Filter = (img: HTMLImageElement) => boolean

export type FindIndex = <T>(
    array: T[],
    predicate: (item: T, idx: number) => boolean
) => number

export type Closest = (
    el: Node | null,
    predicate: (el: HTMLElement) => boolean
) => HTMLElement | false

export type Get = <T>(
    context: Record<string, any>,
    path: string,
    defaultValue: T
) => T

export type Single = <T>(fn: Function) => (...args: any[]) => T

export type BindEvent = (pswp: Pswp, context?: Vue) => void

export type CreatePhotoSwipe = (arg: CreatePhotoSwipeArgs) => Pswp

export type HandleWithoutSize = (pswp: Pswp) => void

export type GetContainSize = (
    areaWidth: number,
    areaHeight: number,
    width: number,
    height: number
) => Size

export type RotateDirection = 'left' | 'right'

export type StyleKey = Exclude<keyof CSSStyleDeclaration, number | 'length' | 'parentRule'>

export type ManualOpenItem = ManualImgItem | ManualHtmlItem
