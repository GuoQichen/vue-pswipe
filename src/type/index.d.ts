import PhotoSwipeUI_Default from 'photoswipe/dist/photoswipe-ui-default'

export interface Options extends PhotoSwipeUI_Default.Options {
    galleryPIDs?: number
}

export type Filter = (img: HTMLImageElement) => boolean
