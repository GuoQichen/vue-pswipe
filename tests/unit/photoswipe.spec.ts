/* eslint-disable import/first */
import {
    PhotoSwipe,
    PhotoSwipeMock,
    createPswp,
    createPswpUI,
    fakeSrc,
    mockImageOnload,
    getFakeImages,
} from './util'

import { BeforeOpenEvent, BeforeOpen } from '@/type'

describe('photoswipe.vue', () => {
    beforeEach(() => {
        expect(PhotoSwipe).not.toHaveBeenCalled()
    })

    afterEach(() => {
        PhotoSwipe.mockClear()
    })

    it('render Photoswipe', () => {
        expect(createPswp({
            withClick: false,
        })).toMatchSnapshot()
    })

    describe(':props', () => {
        describe(':options', () => {
            const createOptionsPswp = (options: Record<string, any> = {}) => {
                createPswp({
                    propsData: {
                        options,
                    },
                })

                return PhotoSwipeMock.getReceiveOptions()
            }

            it('render default options', () => {
                const options = createOptionsPswp()
                expect(JSON.stringify(options, null, 2)).toMatchSnapshot()
            })

            it('set options', () => {
                const bgOpacity = 0.5
                const options = createOptionsPswp({
                    bgOpacity,
                })

                expect(options.bgOpacity).toBe(bgOpacity)
            })
        })

        describe(':auto', () => {
            const createAutoPswp = (auto: boolean) =>
                createPswp({
                    defaultSlots: `<img src="${fakeSrc}" />`,
                    propsData: {
                        auto,
                    },
                })

            it('without auto', () => {
                createAutoPswp(false)
                expect(PhotoSwipe).not.toBeCalled()
            })

            it('with auto', () => {
                createAutoPswp(true)
                expect(PhotoSwipe).toBeCalled()
            })
        })

        describe(':bubble', () => {
            const createBubblePswp = (bubble: boolean) => {
                createPswp({
                    defaultSlots:
                    `
                        <div v-pswp="'${fakeSrc}'">
                            <img src="${fakeSrc}" />
                        </div>
                    `,
                    propsData: {
                        bubble,
                    },
                })
            }

            it('without bubble', () => {
                createBubblePswp(false)
                expect(PhotoSwipe).not.toBeCalled()
            })

            it('with bubble', () => {
                createBubblePswp(true)
                expect(PhotoSwipe).toBeCalled()
            })
        })

        describe(':rotate', () => {
            it('set pswpUI rotate', () => {
                const pswpUIWrapper = createPswpUI()

                createPswp({
                    propsData: {
                        rotate: true,
                    },
                })

                expect(pswpUIWrapper.vm.$data.rotate).toBe(true)
            })
        })

        describe(':lazy', () => {
            const createLazyPswp = (lazy: boolean, fakeImageLength: number) => {
                const fakeImages = getFakeImages(fakeImageLength)

                const wrapper = createPswp({
                    withClick: false,
                    defaultSlots: fakeImages.map((src, index) => ({
                        data: () => ({
                            pswpItem: {
                                src,
                            },
                        }),
                        template:
                        `
                            <img
                                id="img-${index}"
                                :src="pswpItem.src"
                                v-pswp="pswpItem"
                            />
                        `,
                    })),
                    propsData: {
                        lazy,
                    },
                })

                return {
                    wrapper,
                    fakeImages,
                }
            }

            beforeEach(() => {
                mockImageOnload.enable()
            })

            afterEach(() => {
                mockImageOnload.disable()
            })

            it('only load current, the previous and the next', () => {
                const len = 10
                const { wrapper, fakeImages } = createLazyPswp(true, len)

                wrapper.find('#img-5').trigger('click')
                const shouldLoadedIndex = [4, 5, 6]

                expect(mockImageOnload.isLoaded(
                    shouldLoadedIndex.map(i => fakeImages[i]),
                )).toBe(true)

                expect(mockImageOnload.isLoaded(
                    [...Array(len).keys()]
                        .filter(i => !shouldLoadedIndex.includes(i))
                        .map(i => fakeImages[i]),
                )).toBe(false)
            })

            it('preload all image', () => {
                const len = 10
                const { wrapper, fakeImages } = createLazyPswp(false, len)

                wrapper.find('#img-5').trigger('click')

                expect(mockImageOnload.isLoaded(
                    [...Array(len).keys()].map(i => fakeImages[i]),
                )).toBe(true)
            })
        })
    })

    describe('@event', () => {
        describe('@beforeOpen', () => {
            const getMock = (continued: boolean) => jest.fn(
                (target: BeforeOpenEvent, next: BeforeOpen) => {
                    next(continued)

                    continued
                        ? expect(PhotoSwipe).toBeCalled()
                        : expect(PhotoSwipe).not.toBeCalled()
                },
            )

            const createBeforeOpenPswp = (continued: boolean) => {
                const mock = getMock(continued)

                createPswp({
                    listeners: {
                        beforeOpen: mock,
                    },
                })

                return {
                    mock,
                }
            }

            it('called with specified parameters', () => {
                const { mock } = createBeforeOpenPswp(true)
                expect(mock.mock.calls[0].length).toBe(2)
            })

            it('abort open PhotoSwipe', () => {
                const { mock } = createBeforeOpenPswp(false)
                expect(mock).toBeCalled()
            })
        })

        describe('@opened', () => {
            it('get pswp from cb', () => {
                const mock = jest.fn()
                const wrapper = createPswp({
                    withClick: false,
                })
                wrapper.vm.$on('opened', mock)

                wrapper.find('img').trigger('click')
                const pswp = PhotoSwipeMock.getPswp()

                expect(mock).toBeCalled()
                expect(mock).toBeCalledWith(pswp)
            })
        })
    })
})
