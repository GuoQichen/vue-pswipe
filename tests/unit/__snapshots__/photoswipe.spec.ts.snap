// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`photoswipe.vue :props :options render default options 1`] = `
"{
  \\"galleryUID\\": 0,
  \\"history\\": false,
  \\"zoomEl\\": true,
  \\"shareEl\\": true,
  \\"shareButtons\\": [
    {
      \\"id\\": \\"download\\",
      \\"label\\": \\"Download image\\",
      \\"url\\": \\"{{raw_image_url}}\\",
      \\"download\\": true
    }
  ],
  \\"index\\": 0,
  \\"showHideOpacity\\": false
}"
`;

exports[`photoswipe.vue render Photoswipe 1`] = `<div class="pswipe-gallery" data-pswp-uid="0"><img src="https://placeimg.com/640/480/any" data-pswp-src="https://placeimg.com/640/480/any"></div>`;
